### PR TITLE
Iox #1150 remove findservice with container from discovery

### LINKED ISSUE #1150: Remove findService with container return value from Discovery
## Brief feature description

- remove `ServiceContainer findService(...)` from Discovery
- adapt tests accordingly
- semi-related: store ServiceRegistry member of ServiceDiscovery with dynamic memory (`unique_ptr`) instead of the stack (obtained from a Sample)

## Rationale

Reduce stack consumption of ServiceDiscovery

### PR DESCRIPTION
**Pull request by MatthiasKillat**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1150

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 1]
```diff
@@ -68,15 +70,17 @@ class ServiceDiscovery
     void invalidateTrigger(const uint64_t uniqueTriggerId);
     iox::popo::WaitSetIsConditionSatisfiedCallback
     getCallbackForIsStateConditionSatisfied(const popo::SubscriberState state);
-    roudi::ServiceRegistry m_serviceRegistry;
+
+    //  use dynamic memory to reduce stack usage,
+    /// @todo improve solution to avoid stack usage without using dynamic memory
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:75]: Please add a issue number here `/// @todo #123` and create the respective issue to clean this up.

Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:75]: Yes sure, was a placeholder.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 2]
```diff
@@ -665,8 +635,8 @@ TYPED_TEST(ServiceRegistry_test, SearchInFullRegistryWorks)
     // It could also be achieved with determinstic string enumeration instead of randomness,
     // but it is more cumbersome and not required here.
 
-    this->sut->find(searchResult, service, instance, event);
-    ASSERT_EQ(searchResult.size(), 1);
+    this->find(service, instance, event);
+    ASSERT_EQ(this->searchResult.size(), 1);
 }
 
 using Entry = iox::roudi::ServiceRegistry::ServiceDescriptionEntry;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:642]: This can be removed.

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 3]
```diff
@@ -33,7 +35,7 @@ enum class ServiceDiscoveryEvent : popo::EventEnumIdentifier
 class ServiceDiscovery
 {
   public:
-    ServiceDiscovery() noexcept = default;
+    ServiceDiscovery();
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:38]: Please remove the `findService` method that returns a `ServiceContainer`.

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 4]
```diff
@@ -68,15 +70,17 @@ class ServiceDiscovery
     void invalidateTrigger(const uint64_t uniqueTriggerId);
     iox::popo::WaitSetIsConditionSatisfiedCallback
     getCallbackForIsStateConditionSatisfied(const popo::SubscriberState state);
-    roudi::ServiceRegistry m_serviceRegistry;
+
+    //  use dynamic memory to reduce stack usage,
+    /// @todo improve solution to avoid stack usage without using dynamic memory
+    std::unique_ptr<roudi::ServiceRegistry> m_serviceRegistry;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:76]:
```suggestion
    std::unique_ptr<roudi::ServiceRegistry> m_serviceRegistry{new roudi::ServiceRegistry};
```

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 5]
```diff
@@ -188,47 +159,26 @@ TEST_F(ServiceDiscovery_test, OfferExistingServiceMultipleTimesIsRedundant)
...
-    compareAndClearServiceContainers(serviceContainer, searchResultOfFindServiceWithFindHandler);
-
-    serviceContainer = sut.findService(SERVICE_DESCRIPTION.getServiceIDString(),
-                                       SERVICE_DESCRIPTION.getInstanceIDString(),
-                                       SERVICE_DESCRIPTION.getEventIDString());
+    findService(SERVICE_DESCRIPTION.getServiceIDString(),
+                SERVICE_DESCRIPTION.getInstanceIDString(),
+                SERVICE_DESCRIPTION.getEventIDString());
     ASSERT_THAT(serviceContainer.size(), Eq(1U));
     EXPECT_THAT(*serviceContainer.begin(), Eq(SERVICE_DESCRIPTION));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:181]: Please add again an additional `findService` call to fulfill the test intention.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 6]
```diff
@@ -319,14 +249,13 @@ TEST_F(ServiceDiscovery_test, FindServiceWithInstanceAndEventWildcardReturnsAllM
...
+    findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard);
     ASSERT_THAT(serviceContainer.size(), Eq(3U));
     EXPECT_TRUE(serviceContainer == serviceContainerExp);
-
-    sut.findService(SERVICE, iox::capro::Wildcard, iox::capro::Wildcard, findHandler);
-    EXPECT_TRUE(searchResultOfFindServiceWithFindHandler == serviceContainerExp);
 }
 
+#if 1
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:258]:
```suggestion
```

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 7]
```diff
@@ -744,5 +621,6 @@ TEST_F(ServiceDiscovery_test, ServiceDiscoveryNotifiedbyListenerFindsSingleServi
     ASSERT_FALSE(serviceContainer.empty());
     EXPECT_THAT(serviceContainer.front(), Eq(serviceDescriptionToSearchFor));
 }
+#endif
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:624]:
```suggestion
```

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 8]
```diff
@@ -21,16 +21,13 @@ namespace iox
...
+void ServiceDiscovery::update()
 {
-    ServiceContainer searchResult;
-
-    auto lambda = [&](const capro::ServiceDescription& entry) { searchResult.emplace_back(entry); };
-    findService(service, instance, event, lambda);
-
-    return searchResult;
+    // allows us to use update and hence findService concurrently
+    std::lock_guard<std::mutex> lock(m_serviceRegistryMutex);
```
Comment (mossmaurice) [inline iceoryx_posh/source/runtime/service_discovery.cpp:27]: YAGNI? Currently `update()` is private and can't be called concurrently by the user. Please either make `update()` public or remove the lock. I would prefer the latter.

Comment (MatthiasKillat) [reply iceoryx_posh/source/runtime/service_discovery.cpp:27]: No, clean design. And stop this YAGNI please.. `update` exists to cleanly separate findService and update. `findService` is calling update. Only the update needs to be concurrently protected (because it writes), not the whole findService (as the remainder only reads). -> Keep critical sections small

I do not want to have bug reports when it is used in multiple threads and the mutex prevents problems here.

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

**Comment by FerdinandSpitzschnueffler** [comment]
@MatthiasKillat Can you please check if we have to remove code examples for the removed `findService` method from the `iceoryx-v2-0-0.md?

**Comment by MatthiasKillat** [comment]
@FerdinandSpitzschnueffler 

> Can you please check if we have to remove code examples for the removed findService method from the `iceoryx-v2-0-0.md?

Yes, I will do so and change it in the upcoming example PR https://github.com/eclipse-iceoryx/iceoryx/pull/1117.

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 8 above)
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp:27]: @MatthiasKillat Is the use case of concurrent `findService` so prominent? Several other parts of the user API is not usable concurrently. In the past we added a comment or doxygen hint for these parts. Why not just add a doxygen comment here too? The user can always use a lock around e.g. with `concurrent::smart_lock`

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 9]
```diff
@@ -84,26 +82,19 @@ class ServiceRegistry
...
-              const cxx::optional<capro::IdString_t>& service,
-              const cxx::optional<capro::IdString_t>& instance,
-              const cxx::optional<capro::IdString_t>& event) const noexcept;
-
-    /// @copydoc ServiceDiscovery::findService
+    /// @param[in] callable, callable to apply to each matching entry
     void find(const cxx::optional<capro::IdString_t>& service,
               const cxx::optional<capro::IdString_t>& instance,
               const cxx::optional<capro::IdString_t>& event,
               cxx::function_ref<void(const ServiceDescriptionEntry&)> callable) const noexcept;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:92]:
```suggestion
              cxx::function_ref<void(const ServiceDescriptionEntry&)> callableForEach) const noexcept;
```

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 10]
```diff
@@ -84,26 +82,19 @@ class ServiceRegistry
...
 
-    /// @todo #415 this may not be needed later or we can move applyToAll to the public interface,
-    ///       (we want to avoid large containers on the stack)
-    /// @brief Returns all service descriptions as copy
-    /// @return ServiceDescriptionVector_t, copy of complete service registry
-    const ServiceDescriptionVector_t getServices() const noexcept;
+    /// @brief Applys a callable to all entries
+    /// @param[in] callable, callable to apply to each entry
+    /// @note Can be used to obtain all entries or count them
+    void applyToAll(cxx::function_ref<void(const ServiceDescriptionEntry&)> callable) const noexcept;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:97]:
```suggestion
    void applyToAll(cxx::function_ref<void(const ServiceDescriptionEntry&)> callableForEach) const noexcept;
```

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 11]
```diff
@@ -41,24 +44,13 @@ enum class ServiceDiscoveryEvent : popo::EventEnumIdentifier
 class ServiceDiscovery
 {
   public:
-    ServiceDiscovery() noexcept = default;
+    ServiceDiscovery() = default;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:47]: Why did you remove `noexcept`? IMHO we should not do this now but as part of #1032

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp [hunk 12]
```diff
@@ -78,15 +70,20 @@ class ServiceDiscovery
     void invalidateTrigger(const uint64_t uniqueTriggerId);
     iox::popo::WaitSetIsConditionSatisfiedCallback
     getCallbackForIsStateConditionSatisfied(const popo::SubscriberState state);
-    roudi::ServiceRegistry m_serviceRegistry;
+
+    // use dynamic memory to reduce stack usage
+    /// @todo #1155 improve solution to avoid stack usage without using dynamic memory
+    std::unique_ptr<roudi::ServiceRegistry> m_serviceRegistry{new iox::roudi::ServiceRegistry};
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:76]: Nitpick: Once `m_deleter` is a `cxx::function` we could use the `cxx::unique_ptr` here and reduce the dependency to the STL. Not necessary as it's a temporary solution anyway.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 13]
```diff
@@ -93,40 +98,35 @@ class ServiceDiscoveryPubSub_test : public RouDi_GTest
...
-{
-    const auto size = lhs.size();
-    ASSERT_THAT(size, Eq(rhs.size()));
-    for (auto i = 0U; i < size; i++)
+    void findService(const optional<IdString_t>& service,
+                     const optional<IdString_t>& instance,
+                     const optional<IdString_t>& event) noexcept
     {
-        EXPECT_THAT(lhs[i], Eq(rhs[i]));
+        ServiceDiscoveryPubSub_test::findService(service, instance, event, CommunicationKind::KIND);
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:127]: Nice!

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 14]
```diff
@@ -704,8 +505,14 @@ TYPED_TEST(ServiceDiscovery_test, FindServiceReturnsMaxProducerServices)
 
     // we want to check whether we find all the services, including internal ones like introspection,
     // so we first search for them without creating other services to obtain the internal ones
-    ServiceContainer serviceContainerExp = this->sut.findService(
-        iox::capro::Wildcard, iox::capro::Wildcard, iox::capro::Wildcard, TestFixture::CommunicationKind::KIND);
+
+    ServiceContainer serviceContainerExp;
+    auto findInternalServices = [&](const ServiceDescription& s) { serviceContainerExp.emplace_back(s); };
+    this->sut.findService(iox::capro::Wildcard,
```
Comment (mossmaurice) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:511]: I think you could use `FerviceDiscoveryPubSub_test::findService()` here as well.

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 8 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/runtime/service_discovery.cpp:27]: @mossmaurice 
The user can, but she cannot do as fine-grained as we can as she cannot access the internals (the update part). 

I would think it is a default use case for the following reason: I would use this `Servicediscovery` (or a custom one build on this) as a singleton, i.e. only one per application (either by using static or with some global). There is no point in creating multiple ones, since all they do is hog up memory. If I now have multiple threads I can easily use the ServiceDiscovery from each of them without worries as the `findService` call is protected by mutex.

It is different from e.g. subscribers which have a theoretically smaller scope and exist multiple times

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 15]
```diff
@@ -518,52 +531,10 @@ TYPED_TEST(ServiceRegistry_test, AddingMultipleServiceDescriptionAndRemovingAllD
     this->sut.remove(service2);
     this->sut.remove(service3);
 
-    this->sut->find(this->searchResult, iox::capro::IdString_t("a"), iox::capro::Wildcard, iox::capro::Wildcard);
+    this->find(iox::capro::IdString_t("a"), iox::capro::Wildcard, iox::capro::Wildcard);
     EXPECT_THAT(this->searchResult.size(), Eq(0));
 }
 
-TYPED_TEST(ServiceRegistry_test, AddingVariousServiceDescriptionAndGetServicesDoesNotReturnDuplicate)
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:525]: I know this is an old test and needs to be reworked. However, I think we still miss a `E` test case from [`ZOMBIES`](https://iceoryx.io/v1.0.1/advanced/best-practice-for-testing/): Exercise Exceptional Behavior. What do you think about creating a complex and random scenario?

**Review by mossmaurice — COMMENTED** [review]
Looks good in general, only minor findings.

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp (hunk 12 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:76]: Yes, but as it is now I would prefer `std::unique_ptr` which properly deals with dynamic memory.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 14 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:511]: I need a specific lambda here with `serviceContainerExp` to collect the internal services.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp (hunk 15 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:525]: Not in this PR, this is what we should consider in https://github.com/eclipse-iceoryx/iceoryx/issues/1065 .

File: iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp (hunk 11 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/runtime/service_discovery.hpp:47]: It will not build in the tests for some GTest reason ...

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 10 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:97]: adds no value IMO

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 14 above)
Comment (mossmaurice) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:511]: You could use the container from the fixture.

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 8 above)
Comment (mossmaurice) [reply iceoryx_posh/source/runtime/service_discovery.cpp:27]: Ok, you convinced me, looking at the `PoshRuntime`:

```
bool PoshRuntimeImpl::sendRequestToRouDi(const IpcMessage& msg, IpcMessage& answer) noexcept
{
    // runtime must be thread safe
    std::lock_guard<posix::mutex> g(m_appIpcRequestMutex);
    return m_ipcChannelInterface.sendRequestToRouDi(msg, answer);
}
```

This means that all `getMiddlewareFoo()` methods are thread-safe.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp (hunk 15 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:525]: @mossmaurice Ah, no then we should not add it in the issue (only tests for `ServiceDiscovery`). I would have to think about what this complex scenario would be but I think we have more urgent matters right now. Definitely in this PR and I would not say that something obvious/critical is missing in these test cases right now, so I would create no issue.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 14 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:511]: No, I need this for comparison ...

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by mossmaurice — APPROVED** [review]
